Automattic/wp-calypso
# Domains: Transfer: Implement Domain Transfers

This PR implements the Domain Transfer feature. It's a bit of a rewrite of @rads 's work since we needed to simplify the flow.

Props @breezyskies for the designs.

### Domain States
As far as transferred are concerned, a domain can be in many states:
1. Locked
2. Unlocked
3. Pending Transfer
4. Prohibited
5. Verification Required

I'll try to simplify this as to the best of my abilities.

#### Locked State
![image](https://cloud.githubusercontent.com/assets/991022/12213801/3b9c32ec-b634-11e5-802d-1ef61ccfcabe.png)
In this state, the domain is locked. You cannot transfer domain out, you cannot request a transfer code. You can use Privacy Protection.

#### Unlocked State
![image](https://cloud.githubusercontent.com/assets/991022/12213799/21d276be-b634-11e5-8cce-3b93dbb66962.png)

In this state, the domain is unlocked. You can transfer the domain out, you can request a transfer code. You cannot use Privacy Protection.

#### Prohibited
![image](https://cloud.githubusercontent.com/assets/991022/12213812/7f2600a6-b634-11e5-8685-3abf7d07641b.png)

The domain is locked and cannot be unlocked, since the domain is new (registered in 60 days), or it has been transferred in the last 60 days. This is enforced by ICANN and cannot be waived.

#### Verification Required
![image](https://cloud.githubusercontent.com/assets/991022/12213815/8d665f44-b634-11e5-8b01-b9a53e9febbe.png)
The domain is locked, and you have failed to confirm your email address with ICANN. In order to be able to unlock your domain, you need to confirm your email address in your WHOIS record.


This is all too complicated, in order to make this process simpler, we batch multiple actions together.

##### Request transfer code
Visible in Locked state. Unlocks the domain, disables privacy, requests a transfer code.

##### Cancel transfer
Visible in Unlocked state. Locks the domain, re-enables privacy if possible. Denies any pending transfers.


Testing
-
### Prohibited
1. Go to Domains -> (fairly new domain) -> Transfer Domain
2. See prohibited Message

### Verification Required
1. Go to Domains -> (unverified domain) -> Transfer Domain
2. See verification required message
3. Click "Resend" button
4. Confirm receipt of email
5. Click link on email
6. Confirm the verification required message is gone

### Normal
1. Go to Domains -> (eligible domain) -> Transfer Domain
2. Click "Request transfer code"
3. Confirm receipt of 2 emails (status change notification and the code itself)
4. Click "Resend confirmation code"
5. Confirm receipt of second confirmation code
6. Click "Cancel Transfer"
7. Confirm it's back to initial state
8. Confirm receipt of status change notification email

Since it might not be possible to test all cases in a single account ( you might not have the domains), if you were able to test a single case; please check below:

- [x] Prohibited
- [x] Verification Required

/cc: @klimeryk @breezyskies (for design review)

ps: Brie, can you check the error / success messages to see if there's anything to be simplified?

pps: This fixes #249
